File: python/python_asyncflow/time_frame.py
"""Time management for AsyncFlow"""

import time
from typing import Optional


class TimeFrame:
    """Manages timing information for the flow system"""
    
    def __init__(self):
        self.now = time.time()
        self.last = self.now
        self.delta = 0.0
    
    def update(self) -> None:
        """Update time frame with current time"""
        now = time.time()
        self.last = self.now
        self.delta = now - self.now
        self.now = now
    
    def update_with_delta(self, delta: float) -> None:
        """Update time frame with specified delta"""
        self.last = self.now
        self.delta = delta
        self.now = self.last + delta

File: src/time_frame.rs
use std::time::{Duration, Instant};

#[derive(Debug, Clone)]
pub struct TimeFrame {
    pub now: Instant,
    pub last: Instant,
    pub delta: Duration,
}

impl TimeFrame {
    pub fn new() -> Self {
        let now = Instant::now();
        Self {
            now,
            last: now,
            delta: Duration::ZERO,
        }
    }

    pub fn update(&mut self) {
        let now = Instant::now();
        self.last = self.now;
        self.delta = now.duration_since(self.now);
        self.now = now;
    }

    pub fn update_with_delta(&mut self, delta: Duration) {
        self.last = self.now;
        self.delta = delta;
        self.now = self.last + delta;
    }
}

impl Default for TimeFrame {
    fn default() -> Self {
        Self::new()
    }
}